### PR TITLE
Fixed all of the array patterns and gfx macros

### PR DESCRIPTION
**Pull request by kvark**
An attempt to make it run on rust-beta. Please take it from here, if you can.

cc @mattico @eddyb

### CODE REVIEW

**Comment by eddyb** [comment]
Removing array patterns makes me sad :(.

**Comment by kvark** [comment]
@eddyb you think I'm happy? Going through all of those wasn't the most pleasuring evening for me. Really though, I think this code just needs to use a math library (`vecmath`?). Also, I didn't know about `array.x()` syntax, and I don't think it makes much sense, considering that element semantics might be anything (not just the coordinates).

File: src/minecraft/block_state.rs [hunk 1]
```diff
@@ -580,8 +589,8 @@ pub fn fill_buffer<R: gfx::Resources>(block_states: &BlockStates<R>,
                     });
 
                     // Split the clockwise quad into two clockwise triangles.
-                    buffer.push_all(&[v[0], v[1], v[2]]);
-                    buffer.push_all(&[v[2], v[3], v[0]]);
+                    buffer.push_all(&[v[0].clone(), v[1].clone(), v[2].clone()]);
```
Comment (eddyb) [inline src/minecraft/block_state.rs]: Huh, what's the element type that requires cloning?

Comment (kvark) [inline src/minecraft/block_state.rs]: `Vertex`. Is it a big deal though? We could have derived `Copy` within `gfx_vertex!` easily.

Comment (eddyb) [inline src/minecraft/block_state.rs]: Oh, I guess that makes sense.
I believe that every type that can be Copy should be Copy, but that ship has sailed. (*sigh* why did people forget about lints?)

So don't mind me.

Comment (kvark) [inline src/minecraft/block_state.rs]: There are often compatibility concerns. Like with `ShaderParam` stuff that was `Copy` at some point. It was fine for as long as texture handles were `Copy`, but then they stopped to be. Hence, `ShaderParam` and its usage needed to change. That's why I prefer to stick with `Clone` unless I'm 100% sure that the data is small and always plain. Case of `Vertex` is really undefined though, since it is kinda plain data, but nowhere in the code did I need to copy it implicitly.

**Comment by kvark** [comment]
This should be ready now. I was able to actually run it.

File: src/minecraft/region.rs [hunk 2]
```diff
@@ -3,7 +3,6 @@ use std::fs::{File, Metadata};
 use std::io;
 use std::path::Path;
 use gfx;
-use libc;
```
Comment (mattico) [inline src/minecraft/region.rs]: Don't need libc for `cfg(not(windows))` but do need it for `cfg(windows)`. Can move it to inside the windows `map_fd()` so that there's no warning for unix builds.

**Comment by kvark** [comment]
@mattico thanks for the review! windows build should be good now.

File: src/minecraft/block_state.rs [hunk 3]
```diff
@@ -404,10 +407,14 @@ pub fn fill_buffer<R: gfx::Resources>(block_states: &BlockStates<R>,
         for z in 0..16_usize {
             for x in 0..16_usize {
                 let at = |dir: [i32; 3]| {
-                    let [dx, dy, dz] = dir.map(|x| x as usize);
-                    let [x, y, z] = [x.wrapping_add(dx), y.wrapping_add(dy), z.wrapping_add(dz)].map(|x| x.wrapping_add(16));
+                    let (dx, dy, dz) = (dir[0] as usize, dir[1] as usize, dir[2] as usize);
```
Comment (eddyb) [inline src/minecraft/block_state.rs]: Why did you remove this use of the `map` method on `[T; 3]`?

File: src/minecraft/block_state.rs [hunk 4]
```diff
@@ -404,10 +407,14 @@ pub fn fill_buffer<R: gfx::Resources>(block_states: &BlockStates<R>,
...
+                    let (x, y, z) = (
+                        x.wrapping_add(dx).wrapping_add(16),
+                        y.wrapping_add(dy).wrapping_add(16),
+                        z.wrapping_add(dz).wrapping_add(16)
+                    );
                     let chunk = chunks[y / 16][z / 16][x / 16];
-                    let [x, y, z] = [x, y, z].map(|x| x % 16);
+                    let (x, y, z) = (x % 16, y % 16, z % 16);
```
Comment (eddyb) [inline src/minecraft/block_state.rs]: And another one, is it actually all the `.map` calls?

File: src/chunk.rs [hunk 5]
```diff
@@ -93,7 +93,7 @@ impl<R: gfx::Resources> ChunkManager<R> {
                     columns.map(
                         |cz| cz.map(
```
Comment (eddyb) [inline src/chunk.rs]: I guess not, since this is also working on arrays.

File: src/minecraft/block_state.rs (hunk 3 above)
Comment (kvark) [inline src/minecraft/block_state.rs]: because it returns `[T; 3]`, and I need a tuple

Comment (eddyb) [inline src/minecraft/block_state.rs]: Derp!
